### PR TITLE
Importing .dly files, text files with multiple missing values and resetting select all checkbox

### LINKED ISSUE #5788: Import file with an odd extension
[LIVING01_test.zip](https://github.com/africanmathsinitiative/R-Instat/files/4599826/LIVING01_test.zip)

This is a small file of the type sent by Zambia Met service.  It has the extension dly.  Currently the importing (using rio) assumes the format from the extension.  So if I change the extension to txt it works well.
But rio gives the option to override the default.  If I include the argument format =",", then it reads fine.  That's a bit odd to me, because this file isn't comma separated, but it works fine from the script window.  
Ideally it would check this earlier, so it can give a preview in our dialogue.

### LINKED ISSUE #5969: Very minor point in File > Open with Excel
This is in importing from Excel.  I had 3 files each with 3 sheets to import.
First works fine.  I select all the sheets and there is a check-box for this.
When I return to the dialogue it is (correctly) the same as before, i.e. with the checkbox ticked and all sheets selected.
Now I choose another Excel file.
There are no sheets selected in this new file - correctly - but the select all is still ticked.  It should clear.

### LINKED ISSUE #5977: Reading text files with multiple missing values
Currently you can only specifying a single missing value when reading text files, as stated by @rdstern in this [comment](https://github.com/africanmathsinitiative/R-Instat/pull/5824#issuecomment-688287543), some text files may have multiple missing values.

The import dialog needs to be able to allow users to specify multiple missing values and read the files correctly.

### PR DESCRIPTION
**Pull request by Patowhiz**
Fixes #5788 
Fixes #5977
Fixes #5969

@africanmathsinitiative/developers this is ready for review.

### CODE REVIEW

**Comment by lloyddewit** [comment]
@shadrackkibet please could you review? thanks

**Comment by lloyddewit** [comment]
@shadrackkibet please could you review? thanks

File: instat/dlgImportDataset.vb [hunk 1]
```diff
@@ -434,7 +434,7 @@ Public Class dlgImportDataset
...
+            dlgOpen.Filter = "All Data files|*.csv;*.txt;*.xls;*.xlsx;*.RDS;*.sav;*.tsv;*.csvy;*.feather;*.psv;*.RData;*.json;*.yml;*.dta;*.dbf;*.arff;*.R;*.sas7bdat;*.xpt;*.mtp;*.rec;*.syd;*.dif;*.ods;*.xml;*.html;*.dly|Comma separated files|*.csv|Text data file|*.txt|Excel files|*.xls;*.xlsx|R Data Structure files|*.RDS|SPSS files|*.sav|Tab separated files|*.tsv|CSV with a YAML metadata header|*.csvy|Feather R/Python interchange format|*.feather|Pipe separates files|*.psv|Saved R objects|*.RData|JSON|*.json|YAML|*.yml|Stata files|*.dta|XBASE database files|*.dbf|Weka Attribute-Relation File Format|*.arff|R syntax object|*.R|SAS Files|*.sas7bdat|SAS XPORT|*.xpt|Minitab Files|*.mtp|Epiinfo Files|*.rec|Systat Files|*.syd|Data Interchange Format|*.dif|OpenDocument Spreadsheet|*.ods|Shallow XML documents|*.xml|Single-table HTML documents|*.html|DLY|*.html|All files|*.*"
```
Comment (shadrackkibet) [inline instat/dlgImportDataset.vb:437]: Is this correct? I see `.DLY` is HTML i.e `|DLY|*.html|`

**Comment by shadrackkibet** [comment]
This is potentially not specific to this PR. Using the attached excel file when I select the second sheet I. e (ANALOG).
[Bomet Rainfall Stations_MONITORING_2019.zip](https://github.com/africanmathsinitiative/R-Instat/files/5443879/Bomet.Rainfall.Stations_MONITORING_2019.zip)
 I get the following exception thrown.
image

**Comment by shadrackkibet** [comment]
Some other sheets trigger this exception as well, not all.

**Comment by shadrackkibet** [comment]
I am also struggling to understand the following. Using the attached sample file
[test1.zip](https://github.com/africanmathsinitiative/R-Instat/files/5443955/test1.zip)
When I impute 1 as the missing value string it doesn't work i.e I see the message in the preview "**cannot import this file with current options....**" and then OK is disabled.  If I impute 2 or 3 into the input textbox it works well. I tried changing the file into excel format and it works well, so perhaps this is specific to `.csv` file. I tried with some other `.csv` files and 1 doesn't work.

File: instat/dlgImportDataset.vb [hunk 2]
```diff
@@ -606,8 +605,14 @@ Public Class dlgImportDataset
...
             grpCSV.Show()
-        ElseIf strFileExt = ".csv" Then
+        ElseIf strFileExt = ".csv" OrElse strFileExt = ".dly" Then
+            'treat dly files as csv. override the extension using format parameter 
             strFileType = "CSV"
+            If strFileExt = ".dly" Then
+                clsImportCSV.AddParameter("format", Chr(34) & "csv" & Chr(34))
+            Else
+                clsImportCSV.RemoveParameterByName("format")
+            End If
```
[the remark this answers is not in the document]
Comment (Patowhiz) [reply instat/dlgImportDataset.vb:615]: No longer necessary.

File: instat/dlgImportDataset.vb [hunk 3]
```diff
@@ -878,6 +888,13 @@ Public Class dlgImportDataset
             clsImportExcel.AddParameter("na", GetMissingValueRString(ucrInputMissingValueStringExcel.GetText()))
         ElseIf strFileType = "CSV" Then
             clsImportCSV.AddParameter("na.strings", GetMissingValueRString(ucrInputMissingValueStringCSV.GetText()))
+        ElseIf strFileType = "TXT" Then
+            'for separator we use the function used for csv
+            If rdoSeparatortext.Checked Then
+                clsImportCSV.AddParameter("na.strings", GetMissingValueRString(ucrInputMissingValueStringCSV.GetText()))
+            Else
+                clsImportFixedWidthText.AddParameter("na", GetMissingValueRString(ucrInputMissingValueStringText.GetText()))
+            End If
```
[the remark this answers is not in the document]
Comment (Patowhiz) [reply instat/dlgImportDataset.vb:897]: Done

File: instat/dlgImportDataset.vb (hunk 1 above)
Comment (Patowhiz) [reply instat/dlgImportDataset.vb]: Changed .html to .dly

**Comment by Patowhiz** [comment]
@shadrackkibet thank you for deeply testing this.
I have made the suggested code changes.

In regards to the following comments;

> This is potentially not specific to this PR. Using the attached excel file when I select the second sheet I. e (ANALOG).
> Bomet Rainfall Stations_MONITORING_2019.zip
> I get the following exception thrown.

Please can you re-test this with my latest updates. It worked without the error on my side. See screenshot below.
works1
works2

> I am also struggling to understand the following. Using the attached sample file
> [test1.zip](https://github.com/africanmathsinitiative/R-Instat/files/5443955/test1.zip)
> When I impute 1 as the missing value string it doesn't work i.e I see the message in the preview "**cannot import this file with current options....**" and then OK is disabled. If I impute 2 or 3 into the input textbox it works well. I tried changing the file into excel format and it works well, so perhaps this is specific to `.csv` file. I tried with some other `.csv` files and 1 doesn't work.

I've traced this down to rio package "bug". It appears to suggest that fread function (which it uses for reading csv files) recognises 1 as a boolean. Thus the error shown below. I tested the package with values 0, TRUE and FALSE, it threw a similar error. @rdstern @shadrackkibet should we just let this be as it is? since its a limitation of the package. I couldn't find a work around to the 4 values(0,1,TRUE,FALSE).  
rio_errror

**Comment by shadrackkibet** [comment]
I was able to trigger the exception again. I suspect you are not able to trigger it because this kind of exception is not enabled in your VS. See below
image

**Review by rdstern — APPROVED** [review]
Great to get this added

**Comment by Patowhiz** [comment]
@shadrackkibet yes I've managed to get the error.
@lloyddewit this is a serious logical error, I wonder why it doesn't break the released version. From the implementation of the subroutine, the worksheet column count could be less than the passed data frame column count. This bug is in clsGridLink.vb and it is even in the master branch(I also tested it in my master after enabling the Null Error Exception Setting in Visual Studio). 
@shadrackkibet I have fixed the bug. You can re-test again.
error_serious

**Comment by Patowhiz** [comment]
@lloyddewit @shadrackkibet what's the status of this.

**Comment by lloyddewit** [comment]
@shadrackkibet Thank you for the peer review, are all your comments now resolved?

**Review by lloyddewit — APPROVED** [review]

**Comment by lloyddewit** [comment]
@rdstern since your last approval, one bug has been fixed, please could you re-approve? thanks

**Review by rdstern — APPROVED** [review]
